contracts/contracts/strategies/ConvexGeneralizedMetaStrategy.sol

```diff
@@ -0,0 +1,155 @@
+import { BaseConvexMetaStrategy } from "./BaseConvexMetaStrategy.sol";
+import { StableMath } from "../utils/StableMath.sol";
+
+contract ConvexGeneralizedMetaStrategy is BaseConvexMetaStrategy {
+    using StableMath for uint256;
+    using SafeERC20 for IERC20;
+
+    /* Take 3pool LP and deposit it to metapool. Take the LP from metapool
+     * and deposit them to Convex.
+     */
+    function _lpDepositAll() internal override {
+        IERC20 threePoolLp = IERC20(pTokenAddress);
+        ICurvePool curvePool = ICurvePool(platformAddress);
+
+        uint256 threePoolLpBalance = threePoolLp.balanceOf(address(this));
+        uint256 curve3PoolVirtualPrice = curvePool.get_virtual_price();
+        uint256 threePoolLpDollarValue = threePoolLpBalance.mulTruncate(
+            curve3PoolVirtualPrice
+        );
+
+        uint256[2] memory _amounts = [0, threePoolLpBalance];
+
+        uint256 metapoolVirtualPrice = metapool.get_virtual_price();
+        /**
+         * First convert all the deposited tokens to dollar values,
+         * then divide by virtual price to convert to metapool LP tokens
+         * and apply the max slippage
+         */
+        uint256 minReceived = threePoolLpDollarValue
+            .divPrecisely(metapoolVirtualPrice)
+            .mulTruncate(uint256(1e18) - MAX_SLIPPAGE);
+
+        uint256 metapoolLp = metapool.add_liquidity(_amounts, minReceived);
+
+        bool success = IConvexDeposits(cvxDepositorAddress).deposit(
+            cvxDepositorPTokenId,
+            metapoolLp,
+            true // Deposit with staking
+        );
+
+        require(success, "Failed to deposit to Convex");
+    }
+
+    /**
+     * Withdraw the specified amount of tokens from the gauge. And use all the resulting tokens
+     * to remove liquidity from metapool
+     * @param num3CrvTokens Number of Convex 3pool LP tokens to withdraw from metapool
+     */
+    function _lpWithdraw(uint256 num3CrvTokens) internal override {
+        ICurvePool curvePool = ICurvePool(platformAddress);
+
+        uint256 gaugeTokens = IRewardStaking(cvxRewardStakerAddress).balanceOf(
+            address(this)
+        );
+        /**
+         * Convert 3crv tokens to metapoolLP tokens. Since we have no use for the other token we are also
+         * removing liquidity in the balanced manner (only removing 3CrvLP tokens)
+         */
+        // slither-disable-next-line divide-before-multiply
+        uint256 estimationRequiredMetapoolLpTokens = curvePool
+            .get_virtual_price()
+            .divPrecisely(metapool.get_virtual_price())
+            .mulTruncate(num3CrvTokens);
+
+        int128 metapool3CrvCoinIndex = int128(
+            _getMetapoolCoinIndex(address(pTokenAddress))
+        );
+
+        // add 10% margin to the calculation of required tokens
+        // slither-disable-next-line divide-before-multiply
+        uint256 estimatedMetapoolLPWithMargin = (estimationRequiredMetapoolLpTokens *
```

Comment:
I'm concerned that this is in effect a hardcoded slippage check. Perhaps we should be using something that runs off our metapool slippage config.

Reply:
This code is altered and has another approach in this PR: https://github.com/OriginProtocol/origin-dollar/pull/1161